### PR TITLE
App bar pin cluster

### LINKED ISSUE #9587: APP Bar - Pinning clusters
**Describe the solution you'd like**

We want to be able to pin a cluster by clicking it from the menu.

## AC:
- Allow the user to pin a cluster by clicking it
- Store this configuration of the saved clusters in the same format as we get it here (https://github.com/rancher/dashboard/blob/master/shell/components/nav/TopLevelMenu.vue#L64) 

Specs: https://xd.adobe.com/view/04a2e3b8-4895-405d-a067-5f2c88e83738-51f1/specs/

### PR DESCRIPTION
**Pull request by bisht-richa**
### Summary
Fixes #9587

- Users can pin a cluster by clicking it from the menu.
- Users can pin a cluster even if it is not ready.

### Screenshot/Video

Screenshot 2023-09-12 at 10 12 24

====================
Screenshot 2023-09-12 at 10 12 39

======== Pnined cluster list ===========
Screenshot 2023-09-12 at 10 12 50

======== Pnined cluster list in close App bar ===========
Screenshot 2023-09-12 at 10 13 03

### CODE REVIEW

File: cypress/e2e/po/side-bars/burger-side-menu.po.ts [hunk 1]
```diff
@@ -93,6 +93,24 @@ export default class BurgerMenuPo extends ComponentPo {
     return this.self().find('.body .clusters .cluster.selector.option');
   }
 
+  clustersList(): Cypress.Chainable {
+    return this.self().find('.body .clustersList .cluster.selector.option');
+  }
+
+  pinCluster(): Cypress.Chainable {
```
Comment (richard-cox) [inline cypress/e2e/po/side-bars/burger-side-menu.po.ts:100]: can this be called `pinFirstCluster`?

File: cypress/e2e/po/side-bars/burger-side-menu.po.ts [hunk 2]
```diff
@@ -93,6 +93,24 @@ export default class BurgerMenuPo extends ComponentPo {
     return this.self().find('.body .clusters .cluster.selector.option');
   }
 
+  clustersList(): Cypress.Chainable {
```
Comment (richard-cox) [inline cypress/e2e/po/side-bars/burger-side-menu.po.ts:96]: this and clusters above are the same ( .body > .clusters > .clustersList )

File: cypress/e2e/po/side-bars/burger-side-menu.po.ts [hunk 3]
```diff
@@ -93,6 +93,24 @@ export default class BurgerMenuPo extends ComponentPo {
...
+    return this.self().find('.body .clustersList .cluster.selector.option');
+  }
+
+  pinCluster(): Cypress.Chainable {
+    return this.clustersList().first().trigger('mouseover').find('.pin')
+      .invoke('show')
+      .click();
+  }
+
+  pinClustersList(): Cypress.Chainable {
```
Comment (richard-cox) [inline cypress/e2e/po/side-bars/burger-side-menu.po.ts:106]: --> `clusterPinnedList`

File: cypress/e2e/po/side-bars/burger-side-menu.po.ts [hunk 4]
```diff
@@ -93,6 +93,24 @@ export default class BurgerMenuPo extends ComponentPo {
...
+    return this.self().find('.body .clusters .clustersPinned');
+  }
+
+  unpinClusters(): Cypress.Chainable {
```
Comment (richard-cox) [inline cypress/e2e/po/side-bars/burger-side-menu.po.ts:110]: this doesn't unpin, it just returns the options of all pinned clusters.

File: cypress/e2e/tests/navigation/side-nav/main-side-menu.spec.ts [hunk 5]
```diff
@@ -24,6 +24,15 @@ describe('Side Menu: main', () => {
     burgerMenuPo.clusters().should('exist');
   });
 
+  it('Pinned and unpinned cluster', { tags: '@adminUser' }, () => {
```
Comment (richard-cox) [inline cypress/e2e/tests/navigation/side-nav/main-side-menu.spec.ts:27]: can we run this one as a standard user as well

File: shell/components/nav/TopLevelMenu.vue [hunk 6]
```diff
@@ -64,6 +67,8 @@ export default {
       return this.clusters.length > this.maxClustersToShow;
     },
 
+    pinnedClusters: mapPref(PINNED_CLUSTERS),
```
Comment (richard-cox) [inline shell/components/nav/TopLevelMenu.vue:70]: no longer used here

File: shell/components/nav/TopLevelMenu.vue [hunk 7]
```diff
@@ -377,36 +400,94 @@ export default {
...
-                  v-else
-                  class="option cluster selector disabled"
+                  <nuxt-link
+                    v-if="c.ready"
+                    :data-testid="`menu-cluster-${ c.id }`"
+                    class="cluster selector option"
+                    :to="{ name: 'c-cluster-explorer', params: { cluster: c.id } }"
+                  >
+                    <ClusterIconMenu
+                      v-tooltip="getTooltipConfig(c?.label)"
```
Comment (richard-cox) [inline shell/components/nav/TopLevelMenu.vue:418]: this was updated in the upstream pr. no `?` needed in `c?.label`

File: shell/components/nav/TopLevelMenu.vue [hunk 8]
```diff
@@ -377,36 +400,94 @@ export default {
...
+                    <Pinned
+                      :cluster="c"
+                    />
+                  </nuxt-link>
+                  <span
+                    v-else
+                    class="option cluster selector disabled"
+                  >
+                    <ClusterIconMenu
+                      v-tooltip="getTooltipConfig(c?.label)"
```
Comment (richard-cox) [inline shell/components/nav/TopLevelMenu.vue:434]: Same here, no `?`

**Review by richard-cox — CHANGES_REQUESTED** [review]
- Still outstanding
  - clusters that are pinned should always show regardless of window size. at the moment they will be collapsed as per non-pinned clusters
- From Sorin's PR. He should have mentioned these, but they might have gotten lost
  - `.side-menu .home svg` should have a `margin-left` of 9px rather than 10px
  - With legacy turned on and an extension installed with with a menu item... there's a blip at about 750px window height where no clusters are visible. With legacy turned off (but still with the extension menu item) it does not happen. I think it was agreed no matter what happens at least two clusters will always be shown


Not For this PR, just tracking...
- discussion - cluster list consuming all available space, not maxed at 10
- discussion - adding more colour / branding given removal of provider icon
- future work - once this one is merged I need to test extensions that use standalone mode

**Review by richard-cox — DISMISSED** [review]
This is good to merge, but can you create an issue to resolve something separately?

If the user pins all clusters the separator line is shown and then the text `No matching clusters`

image

When there's no filter text and all clusters are pinned we just shouldn't show that line and text.

**Comment by bisht-richa** [comment]
The smallIdentifier functionality will be extracted from the component and tested separately as part of https://github.com/rancher/dashboard/issues/9658

File: cypress/e2e/tests/navigation/side-nav/main-side-menu.spec.ts [hunk 9]
```diff
@@ -39,22 +44,6 @@ describe('Side Menu: main', () => {
     BurgerMenuPo.checkTooltipOn();
   });
 
-  it('Check that the local cluster badge matches LCL, custom badge should match EX', { tags: ['@adminUser'] }, () => {
```
Comment (richard-cox) [inline cypress/e2e/tests/navigation/side-nav/main-side-menu.spec.ts:42]: Can you create an issue to bring this back in at some point? It tested 
- abbreviations
- clicking a cluster menu item takes you to a specific cluster
- custom cluster label
so there's a good bit of value there

Comment (richard-cox) [reply cypress/e2e/tests/navigation/side-nav/main-side-menu.spec.ts:42]: It should also run for standardUser, who it think we set up with access to the local cluster

**Review by richard-cox — APPROVED** [review]
Just one comment to create an issue

File: cypress/e2e/tests/navigation/side-nav/main-side-menu.spec.ts (hunk 9 above)
Comment (bisht-richa) [reply cypress/e2e/tests/navigation/side-nav/main-side-menu.spec.ts:42]: @scures has already created one https://github.com/rancher/dashboard/issues/9658

Comment (richard-cox) [reply cypress/e2e/tests/navigation/side-nav/main-side-menu.spec.ts:42]: That's for unit testing the function that provides the abbreviation, this is e2e testing. There's crossover with a couple, but not the middle point

**Pull request merged**